File: backend/app/api/progress.py
import json

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.auth import get_current_user_id
from app.db.session import get_db
from app.models.swing import Swing
from app.schemas.swing import AngleTrend, ProgressPoint, ProgressResponse

router = APIRouter(prefix="/api/users/me", tags=["progress"])


@router.get("/progress", response_model=ProgressResponse)
def get_progress(
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    swings = (
        db.query(Swing)
        .filter(
            Swing.user_id == user_id,
            Swing.status == "complete",
            Swing.overall_score.isnot(None),
        )
        .order_by(Swing.created_at.asc())
        .all()
    )

    scores = [
        ProgressPoint(date=s.created_at, score=s.overall_score)
        for s in swings
        if s.overall_score is not None
    ]

    # Extract angle trends from analysis JSON
    angle_data: dict[str, dict[str, list]] = {}
    tracked_angles = [
        "hip_rotation_deg",
        "shoulder_rotation_deg",
        "spine_angle_deg",
        "weight_transfer_ratio",
    ]

    for s in swings:
        if not s.analysis_json:
            continue
        try:
            analysis = json.loads(s.analysis_json)
        except (json.JSONDecodeError, TypeError):
            continue

        for phase_result in analysis.get("phases_detected", []):
            if phase_result.get("phase") != "impact":
                continue
            angles = phase_result.get("angles", {})
            for angle_name in tracked_angles:
                val = angles.get(angle_name)
                if val is None:
                    continue
                if angle_name not in angle_data:
                    angle_data[angle_name] = {"values": [], "dates": []}
                angle_data[angle_name]["values"].append(val)
                angle_data[angle_name]["dates"].append(s.created_at)

    angle_trends = [
        AngleTrend(angle_name=name, values=data["values"], dates=data["dates"])
        for name, data in angle_data.items()
    ]

    return ProgressResponse(scores=scores, angle_trends=angle_trends)

File: api/internal/handler/progress.go
package handler

import (
	"encoding/json"

	"github.com/gofiber/fiber/v2"
	"gorm.io/gorm"

	"github.com/6cclab/swing-detector/api/internal/middleware"
	"github.com/6cclab/swing-detector/api/internal/models"
)

type ProgressHandler struct {
	db *gorm.DB
}

func NewProgressHandler(db *gorm.DB) *ProgressHandler {
	return &ProgressHandler{db: db}
}

var trackedAngles = []string{
	"hip_rotation_deg",
	"shoulder_rotation_deg",
	"spine_angle_deg",
	"weight_transfer_ratio",
}

func (h *ProgressHandler) GetProgress(c *fiber.Ctx) error {
	userID := middleware.GetUserID(c)

	var swings []models.Swing
	h.db.Where("user_id = ? AND status = ? AND overall_score IS NOT NULL", userID, "complete").
		Order("created_at ASC").
		Find(&swings)

	scores := make([]models.ProgressPoint, 0, len(swings))
	angleData := make(map[string]*models.AngleTrend)

	for _, s := range swings {
		if s.OverallScore != nil {
			scores = append(scores, models.ProgressPoint{
				Date:  s.CreatedAt,
				Score: *s.OverallScore,
			})
		}

		if s.AnalysisJSON == nil {
			continue
		}

		var analysis struct {
			PhasesDetected []struct {
				Phase  string         `json:"phase"`
				Angles map[string]any `json:"angles"`
			} `json:"phases_detected"`
		}
		if err := json.Unmarshal([]byte(*s.AnalysisJSON), &analysis); err != nil {
			continue
		}

		for _, phase := range analysis.PhasesDetected {
			if phase.Phase != "impact" {
				continue
			}
			for _, angleName := range trackedAngles {
				val, ok := phase.Angles[angleName]
				if !ok || val == nil {
					continue
				}
				fval, ok := val.(float64)
				if !ok {
					continue
				}
				trend, exists := angleData[angleName]
				if !exists {
					trend = &models.AngleTrend{AngleName: angleName}
					angleData[angleName] = trend
				}
				trend.Values = append(trend.Values, fval)
				trend.Dates = append(trend.Dates, s.CreatedAt)
			}
		}
	}

	angleTrends := make([]models.AngleTrend, 0, len(angleData))
	for _, trend := range angleData {
		angleTrends = append(angleTrends, *trend)
	}

	return c.JSON(models.ProgressResponse{
		Scores:      scores,
		AngleTrends: angleTrends,
	})
}
